### PR TITLE
feat(gateway): error handling improvements (500, 502, 504)

### PR DESCRIPTION
**Pull request by hacdias**
See https://github.com/ipfs/go-libipfs/pull/182.

Requires:

- [x] https://github.com/ipfs/go-libipfs/pull/187

### CODE REVIEW

File: test/cli/gateway_test.go [hunk 1]
```diff
@@ -166,10 +166,10 @@ func TestGateway(t *testing.T) {
 		t.Parallel()
 		node.IPFS("name", "publish", "--allow-offline", cid)
 
-		t.Run("GET invalid IPNS root returns 400 (Bad Request)", func(t *testing.T) {
+		t.Run("GET invalid IPNS root returns 500 (Internal Server Error)", func(t *testing.T) {
```
Comment (lidel) [inline test/cli/gateway_test.go:169]: This one sgtm, consistent with https://github.com/ipfs/kubo/pull/9589 (`QmInvalid` is not a CID, but technically _could_ be a DNS name, (e.g. something from https://handshake.org/), so we can't say its Bad Request, could be DNS resolution error.)

File: test/cli/gateway_test.go [hunk 2]
```diff
@@ -129,7 +129,7 @@ func TestGateway(t *testing.T) {
 		t.Run("GET IPFS invalid CID returns 400 (Bad Request)", func(t *testing.T) {
 			t.Parallel()
 			resp := client.Get("/ipfs/QmInvalid/pleaseDontAddMe")
-			assert.Equal(t, 400, resp.StatusCode)
+			assert.Equal(t, 500, resp.StatusCode)
```
Comment (lidel) [inline test/cli/gateway_test.go:132]: I think this one needs to be restored to 400 – on `/ipfs/cid` we can always tell when CID is invalid, and that means we always know when it was 400 Bad Request, and not server error.

**Review by lidel — CHANGES_REQUESTED** [review]

File: test/cli/gateway_test.go (hunk 2 above)
Comment (hacdias) [reply test/cli/gateway_test.go]: Yes, requires https://github.com/ipfs/go-path/pull/66.

However, (https://github.com/ipfs/kubo/pull/9660#discussion_r1115178233) will still fail with 400 then. We probably should just create a follow-up issue for this to ensure name resolution errors bubble up as 500 instead. I think InvalidPath makes sense to be 400. Perhaps `path.IsValid` should not fail with hypothetical DNSLink domains? But that could break a lot of existent code.

File: test/cli/gateway_test.go (hunk 1 above)
Comment (hacdias) [reply test/cli/gateway_test.go:169]: I'm slightly confused. This test, with the updates (https://github.com/ipfs/kubo/pull/9660#discussion_r1115664357) fails locally but passes on CI. I wonder if there's some weird distinction between the CI env (Linux) and my laptop (macOS) that makes the resolution fail at different points, raising different errors.

Comment (lidel) [reply test/cli/gateway_test.go:169]: Let's continue in https://github.com/ipfs/go-libipfs/issues/185

**Comment by hacdias** [comment]
@lidel I reused this PR because the previous PR in go-libipfs was already making the tests fail here.

**Review by lidel — APPROVED** [review]
Thanks! Merging to fix CI here and in go-libipfs.

**Pull request merged**